packages/loopring_v3/contracts/impl/BasicDepositContract.sol

```diff
@@ -0,0 +1,182 @@
+/*
+
+  Copyright 2017 Loopring Project Ltd (Loopring Foundation).
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "../iface/IDepositContract.sol";
+import "../iface/ILoopringV3.sol";
+
+import "../lib/AddressUtil.sol";
+import "../lib/ERC20.sol";
+import "../lib/ERC20SafeTransfer.sol";
+import "../lib/MathUint.sol";
+
+
+/// @title Basic implementation of IDepositContract that just stores
+///        all funds without doing anything with them.
+///
+///        Should be able to work with proxy contracts so the contract
+///        can be updated easily (but with great caution and transparency!)
+///        when necessary.
+///
+/// @author Brecht Devos - <brecht@loopring.org>
+contract BasicDepositContract is IDepositContract
+{
+    using AddressUtil       for address;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+
+    event TokenNotOwnedByUsersWithdrawn(
+        address sender,
+        address token,
+        address feeVault,
+        uint    amount
+    );
+
+    address public exchange;
+    ILoopringV3 public loopring;
+
+    // A map from token address to their accumulated balances
+    mapping (address => uint) public tokenBalances;
+
+    modifier onlyWhenUninitialized()
+    {
+        require(exchange == address(0), "INITIALIZED");
+        _;
+    }
+
+    modifier onlyExchange()
+    {
+        require(msg.sender == exchange, "UNAUTHORIZED");
+        _;
+    }
+
+    function initialize(
+        address exchangeAddress,
+        address loopringAddress
+        )
+        external
```

Comment:
I will add them just to make sure, but it's not needed. All `external` functions in the exchange contract are `nonReentrance` and all functions (except one) are `onlyExchange` so they are safe. `withdrawTokenNotOwnedByUsers` is also safe because of the way it works.